### PR TITLE
[ENH] Adding Time Mixingup Contrastive Learning to Self Supervised module

### PR DESCRIPTION
**Pull request by hadifawaz1999**
Adding a state of the art model TimeMCL [1] to populate the self supervised model, after this will probably add a base model to contain the load/save functions

The model takes as input 2 time series, with no informaiton on the labels, augments a third series by mixing up x1 and x2 as x3 = lamda.x1 + (1-lamda).x2 and the model uses a contrastive learning loss to match again the augmented x3 sample to x1 and x2 and discard the rest of the series in the batch

[1] Wickstrøm, Kristoffer, Michael Kampffmeyer, Karl Øyvind Mikalsen, and Robert Jenssen. "Mixing up contrastive learning: Self-supervised representation learning for time series." Pattern Recognition Letters 155 (2022): 54-61.

### CODE REVIEW

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+
+    Parameters
+    ----------
+    alpha : float, default = 0.2
+        The alpha value for the Beta distribution.
+    mixup_temperature : float, default = 0.5
+        The value that controls the logits smoothness.
+    backbone_network : aeon Network, default = None
+        The backbone network used for the SSL model,
+        it can be any network from the aeon.networks
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: This looks very "squished" i.e. it doesnt go all the way to the margin (wrapping too early) which is just adding lots of unnecessary lines to the file.

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+    how much each original contributed. Given two inputs
+    x1 and x2, it creates a new sample
+    x̃ = λ * x1 + (1 - λ) * x2, where λ ∈ [0, 1].
+    By learning from the mixing ratio λ, it builds strong,
+    general features for other tasks.
+
+    Parameters
+    ----------
+    alpha : float, default = 0.2
+        The alpha value for the Beta distribution.
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: I don't think saying alpha is "the alpha value for..." is very descriptive. What does it do in the beta distribution?

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    mixup_temperature : float, default = 0.5
+        The value that controls the logits smoothness.
+    backbone_network : aeon Network, default = None
+        The backbone network used for the SSL model,
+        it can be any network from the aeon.networks
+        module on condition for it's structure to be
+        configured as "encoder", see _config attribute.
+        For TimeMCL, the default network used is
+        FCNNetwork(n_layers=3,
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: Missing "n_filters=[128, 256, 128]" from default implementation that is specified on 208

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 4]
```diff
@@ -0,0 +1,509 @@
...
+        this parameter is discarded.
+    callbacks : keras callback or list of callbacks,
+        default = None
+        The default list of callbacks are set to
+        ModelCheckpoint and ReduceLROnPlateau.
+    batch_size : int, default = 64
+        The number of samples per gradient update.
+    use_mini_batch_size : bool, default = False
+        Whether or not to use the mini batch size formula.
+    n_epochs : int, default = 1000
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py:106]: Missmatch docstring and code default value which is n_epochs = 2000

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+        save_best_model: bool = False,
+        save_last_model: bool = False,
+        save_init_model: bool = False,
+        best_file_name: str = "best_model",
+        last_file_name: str = "last_model",
+        init_file_name: str = "init_model",
+        callbacks: Callback | list[Callback] | None = None,
+        batch_size: int = 64,
+        use_mini_batch_size: bool = False,
+        n_epochs: int = 2000,
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: Miss match docstring

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+        modelcheckpoint callback is used by default,
+        this condition, if True, will prevent the
+        automatic deletion of the best saved model from
+        file and the user can choose the file name.
+    save_last_model : bool, default = False
+        Whether or not to save the last model, last
+        epoch trained, using the base class method
+        save_last_model_to_file.
+    save_init_model : bool, default = False
+        Whether to save the initialization of the  model.
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: double space before "  model"

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+        save_last_model is set to False, this parameter
+        is discarded.
+    init_file_name : str, default = "init_model"
+        The name of the file of the init model, if
+        save_init_model is set to False,
+        this parameter is discarded.
+    callbacks : keras callback or list of callbacks,
+        default = None
+        The default list of callbacks are set to
+        ModelCheckpoint and ReduceLROnPlateau.
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: I think this is fairly ambiguous about what this actually does. Could we have a more detailed explanation?

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+        self.optimizer_ = (
+            tf.keras.optimizers.Adam() if self.optimizer is None else self.optimizer
+        )
+
+        # compile but won't be used
+        model.compile(loss="mse", optimizer=self.optimizer_)
+
+        return model
+
+    def _mixup_loss(self, z1, z2, z_augmented, contrastive_weight):
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py:412]: If I understand this function correctly it returns the mean and not the per-sample losses. Maybe this should be renamed to something that reflects it is the _mixup_mean not the _mixup_loss

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+                n_filters=[128, 256, 128],
+                kernel_size=[7, 5, 3],
+                dilation_rate=[2, 4, 8],
+            )
+        else:
+            raise ValueError(
+                "The parameter backbone_network", "should be an aeon network."
+            )
+
+        X = X.transpose(0, 2, 1)
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py:222]: Could be nice to have some input validation for alpha > 0 and mixup_temperature > 0

File: aeon/transformations/collection/self_supervised/_time_mcl.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+        Parameters
+        ----------
+        input_shape : tuple[int, int]
+            The shape of the data fed into the input layer, should be (m, d).
+
+        Returns
+        -------
+        output : a compiled Keras Model
+        """
+        import numpy as np
```
Comment (chrisholder) [inline aeon/transformations/collection/self_supervised/_time_mcl.py]: numpy doesnt need to be in here

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 1 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: fixed thanks

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 2 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: added an explanation

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 3 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: thanks fixed

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 4 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py:106]: fixed

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 5 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: fixed

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 6 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: fixed

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 7 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: added an explanation

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 8 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py:412]: thanks didnt notice i call reduce mean inside the funciton, i removed it from the inside the function's return and kept it when inside _fit when function is being called, hence kept the "loss" name

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 9 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py:222]: added

File: aeon/transformations/collection/self_supervised/_time_mcl.py (hunk 10 above)
Comment (hadifawaz1999) [reply aeon/transformations/collection/self_supervised/_time_mcl.py]: thanks
